policy/l4: Fix merging of L7 rules

tgraf opened the pull request:
* Allow merging L4 rules with no L7 rules, no parser and no redirect
  port with other L4 rules

* Correctly write back the copy of rules into the hashtable of per
  endpoint rules

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer reviewed (approved):
LGTM, nice find. I think in addition we should revert commit f78a7bd18016 ("tests/k8s: Flush L3/L4 policy before running L7") so that the runtime k8s tests will also highlight this behaviour.

manalibhutiyani reviewed (approved):
Just went through the pkg/policy/rule.go changes. Looks good.

joestringer reviewed (approved):
New version looks good, thanks.

tgraf committed:
policy/l4: Fix merging of L7 rules

* Allow merging L4 rules with no L7 rules, no parser and no redirect
  port with other L4 rules

* Correctly write back the copy of rules into the hashtable of per
  endpoint rules

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
Revert "tests/k8s: Flush L3/L4 policy before running L7"

This reverts commit f78a7bd1801652cf8cfbda435c77659252168322.

Ability to merge L4 policy with and without L7 rules allows to re-enable
this test.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Log errors encountered while resolving L4 policy

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Consolidate capitalization of protocol name

We have been using "TCP" and "tcp" in various places. Add constants and
use "TCP" consistently across the code base. This fixes a L7 policy
rules merging issue when protocol names did not match.

Policies with differing capitalization will automatically be fixed up.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
policy: Only append L7 rules if not identical to existing rule

 * Introduces Equal() and Exists() functions to compile a L7 rules
   list of unique L7 rules

 * Adds unit tests which represent the minikube getting started guide
   policy example

Signed-off-by: Thomas Graf <thomas@cilium.io>

joestringer commented on pkg/policy/api/rule.go:
```diff
@@ -241,6 +241,15 @@ type EgressRule struct {
+type L4Proto string
+
+const (
+	ProtoTCP L4Proto = "tcp"
```
What's the overlap between these and the L4Protocol constants defined in api/v1/models/port.go ?

tgraf replied:
We aren't managing the policy API via swagger yet so this is separate. I would rather not depend on the models file in the policy API for now. We can discuss options on how to consolidate.

joestringer reviewed (commented):
This looks like a nice improvement.

I think that L4PolicyMap.containsAllL3L4() checks for the value "", which we should be sanitizing to "any" with this patch?

joestringer commented on pkg/policy/api/utils.go:
```diff
@@ -0,0 +1,90 @@
+}
+
+// Exists returns true if the HTTP rule already exists in the list of rules
+func (h *PortRuleHTTP) Exists(rules L7Rules) bool {
```
Should this operate on a copy of L7Rules, or just the pointer?

tgraf replied:
A copy is fine here, the rule data is only read.

joestringer reviewed (approved):
LGTM.

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -27,25 +27,25 @@ const (
-func (r Rule) Validate() error {
-	for _, i := range r.Ingress {
-		if err := i.Validate(); err != nil {
+// Sanitize validates a policy rule
```
Maybe: `s/Sanitize validates/Sanitize sanitizes/`?

tgraf replied:
Fixed

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -192,17 +192,16 @@ func (pp PortProtocol) Validate() error {
 
-// Validate CIDR
-func (cidr CIDR) Validate() error {
+// Sanitize CIDR
```
This one is actually Validating, right?

tgraf replied:
marked private. It is validating only for now but we don't know later on. I've only kept the Rule Sanitize() public and have hidden everything so we can rename/change this as we want in the future.

aanm replied:
On that case can you add a pointer to the receiver?

tgraf replied:
It's a string so it can't be a pointer.

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -226,13 +225,13 @@ func (cidr CIDR) Validate() error {
 // valid, and ensuring that all of the exception CIDR prefixes are contained
 // within the allowed CIDR prefix.
-func (c CIDRRule) Validate() error {
+func (c CIDRRule) Sanitize() error {
```
This one is actually Validating, right?

tgraf replied:
marked private. It is validating only for now but we don't know later on. I've only kept the `Rule` `Sanitize()` public and have hidden everything so we can rename/change this as we want in the future.

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -141,44 +141,44 @@ func (kr PortRuleKafka) Validate() error {
-// Validate validates a port policy rule
-func (pr PortRule) Validate() error {
+// Sanitize validates a port policy rule
+func (pr PortRule) Sanitize() error {
```
Should have pointer in the receiver.

tgraf replied:
Fixed

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -141,44 +141,44 @@ func (kr PortRuleKafka) Validate() error {
-// Validate validates L7 rules
-func (pr *L7Rules) Validate() error {
+// Sanitize validates L7 rules
+func (pr *L7Rules) Sanitize() error {
```
Validation, not sanitizing

tgraf replied:
Fixed

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -64,54 +64,54 @@ func (i IngressRule) Validate() error {
 // TODO we need to add support to check
 // wildcard and prefix/suffix later on.
-func (kr PortRuleKafka) Validate() error {
+func (kr PortRuleKafka) Sanitize() error {
```
Validation, not sanitizing

tgraf replied:
marked private

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -64,54 +64,54 @@ func (i IngressRule) Validate() error {
-// Validate validates an egress policy rule
-func (e EgressRule) Validate() error {
+// Sanitize validates an egress policy rule
+func (e EgressRule) Sanitize() error {
```
Add pointer to the receiver.

tgraf replied:
Fixed

aanm commented on pkg/policy/api/rule_validation.go:
```diff
@@ -27,25 +27,25 @@ const (
-// Validate validates an ingress policy rule
-func (i IngressRule) Validate() error {
+// Sanitize validates an ingress policy rule
+func (i IngressRule) Sanitize() error {
```
Add pointer to the receiver.

tgraf replied:
Fixed
